tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -29,84 +39,189 @@
+ * The list of TfwSrvHash implemented as array.
+ *
+ * The array is chosen instead of a linked list to allow easy binary search
+ * implementation and perhaps to improve spatial locality of TfwSrvHash objects.
+ *
+ * The RCU mechanism is used to protect the list from being modified,
+ * while the data is scanned by tfw_sched_hash_get_srv().
+ *
+ * @rcu is needed to make updates possible in an atomic context.
+ * That happens if a server connection is lost, so the disconnect callback is
+ * executed in a softirq context and a TfwServer is deleted from the scheduler.
  */
-static TfwServer *servers[MAX_SERVERS_N];
-static int servers_n;
-static DEFINE_SPINLOCK(servers_write_lock);
+typedef struct {
+	struct rcu_head rcu;
+	size_t		n;
+	TfwSrvHash	srv_hashes[];
+} TfwSrvHashList;
 
-static int
-find_srv_idx(TfwServer *srv)
-{
-	int i;
-	for (i = 0; i < servers_n; ++i) {
-		if (servers[i] == srv)
-			return i;
-	}
+#define TFW_SRV_HASH_LIST_SIZE(n) 					\
+({									\
+	/* We put two NULLs when the list is empty in order		\
+	 * to simplify the algorithm in tfw_sched_hash_get_srv() */	\
+	size_t _nulls = n ? 1 : 2;					\
+	(sizeof(TfwSrvHashList) + ((n) + _nulls) * sizeof(TfwSrvHash));	\
+})
 
-	return -1;
-}
+/**
+ * A NULL-terminated array of TfwSrvHash that stores all servers added to the
+ * scheduler via tfw_sched_hash_add_srv().
+ *
+ * Concurrent RCU updaters must be synchronized with servers_update_lock.
+ */
+static TfwSrvHashList *tfw_srv_hash_list __rcu;
+static DEFINE_SPINLOCK(srv_hashes_update_lock);
 
+/**
+ * Find an appropriate TfwServer for the HTTP request @msg.
+ * The server is chosen based on the hash value of URI/Host fields of the @msg,
+ * so multiple requests to the same resource are mapped to the same server.
+ *
+ * Higest Random Weight hashing method is involved: for each message we
+ * calculate randomized weights as follows: (msg_hash ^ srv_hash), and pick a
+ * server with the highest weight.
+ * That sticks messages with certain Host/URI to certain IP addresses.
+ * A server always receives requests with some URI/Host values bound to it,
+ * and that holds if connection to the server is lost and then re-established,
+ * and if other servers go offline, and even if Tempesta FW is restarted.
+ *
+ * The drawbacks of HRW hashing are:
+ *  - A weak hash function adds unfairness to the load balancing.
+ *    There may be a case when a server pulls all load from all other servers.
+ *    And such condition (if it occurs although it is quite improbable) is
+ *    quite stable: it cannot be fixed by adding/removing servers and restarting
+ *    the Tempesta FW instance.
+ *  - For every HTTP request, we have to scan the list of all servers to find
+ *    a matching one with the highest weight. That adds some overhead.
+ *    Currently the linear search is used. We may switch to binary search, but
+ *    benchmarks are required to prove that it will behave better on the small
+ *    number of servers that we usually have.
```

Comment:
I have the benchmarks: https://github.com/krizhanovsky/NatSys-Lab/blob/master/array_scans.cc

The program is only 136 lines in length, so it requires almost the same time to check the performance case as to write the comment ;) Don't be lazy.